crates/bevy_ecs/src/query/state.rs

```diff
@@ -160,19 +161,44 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
         access: &mut Access<ArchetypeComponentId>,
     ) -> Self {
         let mut state = Self::new_uninitialized(world);
-        for archetype in world.archetypes.iter() {
-            // SAFETY: The state was just initialized from the `world` above, and the archetypes being added
-            // come directly from the same world.
+        state.update_archetypes_with(world, |state, archetype| {
+            // SAFETY: The validate_world call ensures that the world is the same the QueryState
+            // was initialized from.
             unsafe {
                 if state.new_archetype_internal(archetype) {
                     state.update_archetype_component_access(archetype, access);
                 }
             }
-        }
-        state.archetype_generation = world.archetypes.generation();
+        });
         state
     }
 
+    /// Get a list of archetypes that could potentially be matched by this query,
+    /// if at least one component is required.
+    ///
+    /// We iterate through the required components, and return the smallest list of archetypes
+    /// corresponding to a component.
+    fn get_potential_archetypes(&self, archetypes: &Archetypes) -> Vec<ArchetypeId> {
```

Comment:
Can we return `impl Iterator` here to avoid the allocation?

Reply:
the problem I had was that `impl Iterator + '_` still uses the state's lifetime, which conflicts with accessing the state in 
```
if state.new_archetype_internal(archetype) {
    state.update_archetype_component_access(archetype, access);
}
```

I can probably make it work with split-borrows but it might be a bit ugly